include/posets/downsets/st_backed.hh

```diff
@@ -0,0 +1,95 @@
+#pragma once
+
+#include <cassert>
+#include <iostream>
+#include <cmath>
+#include <memory>
+#include <set>
+#include <vector>
+
+#include <posets/concepts.hh>
+#include <posets/utils/sforest.hh>
+
+namespace posets::downsets {
+  // Forward definition for the operator<<s.
+  template <Vector>
+  class st_backed;
+
+  template <Vector V>
+  std::ostream& operator<<(std::ostream& os, const st_backed<V>& f);
+
+  template <Vector V> class st_backed {
+  private:
+    template <Vector V2>
+    friend std::ostream& operator<<(std::ostream& os, const st_backed<V2>& f);
+
+    int k;
+    size_t dim;
+
+    size_t root{};
+
+  public:
+    typedef V value_type;
+
+    static std::unique_ptr<utils::sforest<V>> forest;
+    static utils::sforest<V>* get_sforest() {
+      return st_backed::forest.get();
+    }
+
+    st_backed() = delete;
+
+    st_backed (std::vector<V>&& elements, unsigned k=0) 
+    {
+      if(!st_backed::forest) {
+        st_backed::forest = std::make_unique<utils::sforest<V>>(k, elements.begin()->size());
```

Comment:
- Follow coding style, spaces before opening parenthesis (here and everywhere).
- It's unclear to me why all st would share the same `k`.